### DISCUSSION TITLE
Maintaining a development environment

### DISCUSSION THREAD

**Discussion by purva-thakre** [unitaryfund/mitiq#2233] [General]
How does everyone here use Jupyter notebook in their development environment?

The discussion in today's Community Call compelled me to create a new virtual environment for the newer dependencies. However, there is an incompatibility when I try to install Jupyter notebook. 

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
qcs-api-client 0.21.6 requires attrs<22.0.0,>=21.3.0, but you have attrs 23.2.0 which is incompatible.
qcs-api-client 0.21.6 requires httpx<0.24.0,>=0.23.0, but you have httpx 0.27.0 which is incompatible.
```
I have had this happen in the past as well. https://github.com/unitaryfund/mitiq/issues/2145#issuecomment-1874582003

Same as last time, I had to use the following to be able to use Jupyter notebook. 

```
pip uninstall traitlets
pip install traitlets==5.13.0
pip install notebook
```

 I am curious to know if there's a better workaround to the issue I tend to run into besides listing Jupyter notebook as a dev dependency. 
 
 Edited to add: This workaround is not ideal as it causes other additional issues https://discord.com/channels/764231928676089909/1220012039494308011

**Comment by natestemen** [comment]
I'm seeing the same error when setting up a fresh environment (i.e. running `conda create -n purva python=3.11`, `conda activate purva`, `conda install pip`, `make install` and finally `pip install notebook`). After this, `notebook` is installed, but fails to properly load the server it starts on my machine when running `jupyter notebook`. I'm also seeing an error when trying to install `jupyterlab`.

One potential solution is to install a global version of jupyter notebook and use that instead of a one inside your mitiq virtual environment. If you try this, let me know if it works!

Personally, I use the [Jupyter extension for vscode](https://marketplace.visualstudio.com/items?itemName=ms-toolsai.jupyter) and let it handle the installs. I'm not totally sure where it puts the `notebook` download, but I haven't run into any problems like this before. I _think_ this extension only installs `ipykernel`.

**Comment by purva-thakre** [comment]
> One potential solution is to install a global version of jupyter notebook and use that instead of a one inside your mitiq virtual environment. If you try this, let me know if it works!

This works partially. The import paths have to be consistent for the virtual environment and `jupyter notebook` installation. 

In my virtual environment, I can import `mitiq` when I install `jupyter notebook` globally. There are some problems with importing other dependencies like `cirq` and `numpy`. Will have to mess around with putting all the dependencies on path. 

I think the VS Code extension is a better option.

**Comment by gachchami** [comment]
I have used the current 0.34.0 with dev-requirements.txt without some test packages and notebook 6.4.13
https://github.com/gachchami/qmitra/blob/2c1f8addb860eb4441581cd952afab7cd4e6dea3/requirements.txt

**Reply by natestemen** [reply]
Glad you found a workaround! Which test packages did you remove?

**Reply by gachchami** [reply]
I removed these.

But I do not believe they were consequential. I just removed them as I wasnt not going to need them in my project (that consumes mitiq).

```
# Unit tests, coverage, and formatting/style.
pytest==8.0.0
pytest-xdist[psutil]==3.0.2
pytest-cov==4.0.0
flake8==7.0.0
black==22.10
mypy==1.0.0
isort==5.13.2
types-tabulate
```

**Reply by natestemen** [reply]
Yup you definitely won't need those if you aren't contributing to Mitiq! Nice work selecting those for pruning.

From the looks of it (and correct me if I'm wrong here), it looks like you don't need to install an editable version of mitiq (i.e. clone the repo and install via `requirements.txt`/`dev_requirements.txt`). How did you remove the dev dependencies if you are running `pip install -r requirements.txt` with [this requirements file](https://github.com/gachchami/qmitra/blob/2c1f8addb860eb4441581cd952afab7cd4e6dea3/requirements.txt#L110)?

**Reply by gachchami** [reply]
I do not need an editable version of mitiq for now. The requirements file I created was derived from dev_requirements.txt of the released tag. this was to ensure the project team does not install anything that mitiq will not like in the future. 

I MAY have to have an editable version in the future if we encounter any issues/add docs for the UF project we are working on

**Reply by natestemen** [reply]
If you do not need an editable install, there is potentially an easier way forward for you then others in this thread.

1. Start a fresh virtual environment
1. Run something like `pip install mitiq==0.43.0` (with your desired version of mitiq, or no version)
2. Install your other required dependencies (maybe qiskit, pennylane, and/or numpy, etc)
3. run `pip freeze > requirements.txt`

This will put all the versions of dependencies you installed into the `requirements.txt` file. Hence you will not need to choose what dependencies come from each project. Pip will handle all that for you!

**Reply by gachchami** [reply]
Good suggestion.

I did try that route. But I got conflicts with packages that conda installs on default python. specifically with 

`requests-ntlm 1.1.0`

My team also uses qiskit. installing it was causing a whole set of other issues (due to updating to newer versions of qiskit). so I used the dev_requirements.txt to ensure I install all the dependencies that were compatible with mitiq. 

For any other packages I would do `pip install <package> --dry-run` and `pipdeptree` to see what are the dependencies and could there be any conflicts.

**Reply by natestemen** [reply]
> I got conflicts with packages that conda installs on default python

Would you mind walking me through what causes the problems? I use conda to manage environments and haven't encountered this issue before, but we likely have slightly different workflows here. I want to ensure it's smooth for all download processes :)

> qiskit

We don't yet support qiskit 1.0, but hope to in the next release or two!

> `pipdeptree`

cool, TIL!

**Reply by gachchami** [reply]
I tried to reproduce this. But I think I did not start with a clean env. I probably installed qiskit or jupyter before trying to install mitiq. this could have caused the conflicts when I installed mitiq.

**Comment by cosenal** [comment]
Here is a report from my recent trip to dependency hell 🔥 Note that I use Jupyterlab instead of Jupyter Notebook, but the conflicts may be similar.

When installing Jupyterlab with `pip install jupyterlab` in the virtual env, the latest version is installed, that is, `jupyterlab-4.1.5`. That carries as dependencies `httpx` and `attrs`. I used the [pipdeptree](https://pypi.org/project/pipdeptree/) tool to understand conflicting dependencies (same tool recommended by @gachchami above.)

```
* cirq-ft==1.3.0
 - cirq-core [required: ==1.3.0, installed: 1.1.0]
* qcs-api-client==0.21.6
 - attrs [required: >=21.3.0,<22.0.0, installed: 23.2.0]
 - httpx [required: >=0.23.0,<0.24.0, installed: 0.27.0]
```
So far, nothing new from the warning message in @purva-thakre's original post.

The result of `pipdeptree -r -p httpx` is more interesting.
```
httpx==0.27.0
├── jupyterlab==4.1.5 [requires: httpx>=0.25.0]
└── qcs-api-client==0.21.6 [requires: httpx>=0.23.0,<0.24.0] 
```
By looking at the commit history of jupyterlab, It turns out that a dependency on httpx>=0.25.0 was introduced recently, so one could pin `jupyterlab==4.0.12`, which is the last version without a strict dependency on `httpx`. However, that doesn't resolve the conflict with `attrs`, which is more complicated. For that, I tried to use the automatic dependency resolver backed in pip, but it timed out 😬

```
pip._vendor.resolvelib.resolvers.ResolutionTooDeep: 200000
```

In any case, it sounds wrong to retro-fit the version of jupyterlab to please the rest of the dependencies. The right way is to move forward with `qcs-api-client`.  Good news is that recent versions of `qcs-api-client` don't bound attrs version, see [this commit](https://github.com/rigetti/qcs-api-client-python/commit/9233cf291c000b9c72b02f85546bf2816eeda87d).

Here is how mitiq depends on an older version of `qcs-api-client`.
```
qcs-api-client==0.21.6
└── pyquil==3.5.4 [requires: qcs-api-client>=0.21.0,<0.22.0]
    └── cirq-rigetti==1.1.0 [requires: pyquil>=3.2.0]
        └── cirq==1.1.0 [requires: cirq-rigetti==1.1.0]
            └── mitiq==0.35.0.dev0 [requires: cirq>=1.0.0,<1.4.0]
``` 
In other words, `pyquil` is the bottleneck at the moment. I will try to upgrade pyquil, and I will make separate post to report back on the progress.

**Reply by cosenal** [reply]
Tracking pyQuil major dependency bump in https://github.com/unitaryfund/mitiq/issues/2254

**Reply by cosenal** [reply]
And in fact it's cirq, and in turns, cirq-rigetti, which is bounding those dependencies. 
@natestemen @purva-thakre See discussion in https://github.com/quantumlib/Cirq/issues/6500

**Reply by purva-thakre** [reply]
@cosenal Nice job with getting out of the dependency hell! I have always ended up giving up halfway out of frustration. :)

Correct me if I am wrong but from what I understand, one possible solution could be:

- separating out the `cirq` related subpackages in `requirements.txt`. 
- Skip installing `cirq-rigetti` for now
- Temporarily disable all unit tests related to `cirq-rigetti`

`Cirq` becoming compatible with pyQuil v4 won't help us that much until #2254 is closed. It's great that they believe upgrading to v4 will be a quick fix. 

From what I see, this dependency hell also occurs if we install a non-editable version of mitiq, right? Then, perhaps we also need to provide some additional information in the docs that a user should still be able to use the jupyter notebook extension in VS Code.

**Reply by cosenal** [reply]
@purva-thakre Yes, that's one possible solution. However, since Cirq is used for the internal circuit representation, removing `cirq-rigetti` from the dependencies would mean that Mitiq wouldn't support pyQuil at all. 
Good news from my initial investigation on pyQuil 4 is that once Cirq becomes compatible with it, #2254 becomes an easy job for us.

**Comment by purva-thakre** [comment]
Going to go ahead and close this discussion since we know which dependency is causing this and there's an open issue related to it.
